pkg/ipmasq/ipmasq.go

```diff
@@ -76,35 +75,82 @@ type IPMasqAgent struct {
 	nonMasqCIDRsFromConfig map[string]net.IPNet
 	nonMasqCIDRsInMap      map[string]net.IPNet
 	ipMasqMap              IPMasqMap
+	watcher                *fsnotify.Watcher
+	stop                   chan struct{}
 }
 
-// Start starts the "ip-masq-agent" controller which is used to sync the ipmasq
-// BPF maps.
-func Start(configPath string, syncPeriod time.Duration) {
-	start(configPath, syncPeriod, &ipmasq.IPMasqBPFMap{}, controller.NewManager())
+func NewIPMasqAgent(configPath string) (*IPMasqAgent, error) {
+	return newIPMasqAgent(configPath, &ipmasq.IPMasqBPFMap{})
 }
 
-func start(configPath string, syncPeriod time.Duration,
-	ipMasqMap IPMasqMap, manager *controller.Manager) {
+func newIPMasqAgent(configPath string, ipMasqMap IPMasqMap) (*IPMasqAgent, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create fsnotify watcher: %s", err)
+	}
+
+	configDir := filepath.Dir(configPath)
+	// The directory of the config should exist at this time, otherwise
+	// the watcher will fail to add
+	if err := watcher.Add(configDir); err != nil {
+		watcher.Close()
+		return nil, fmt.Errorf("Failed to add %q dir to fsnotify watcher: %s", configDir, err)
+	}
 
 	a := &IPMasqAgent{
 		configPath:        configPath,
 		nonMasqCIDRsInMap: map[string]net.IPNet{},
 		ipMasqMap:         ipMasqMap,
+		watcher:           watcher,
 	}
 
+	return a, nil
+}
+
+// Start starts the ip-masq-agent goroutine which tracks the config file and
+// updates the BPF map accordingly.
+func (a *IPMasqAgent) Start() {
 	if err := a.restore(); err != nil {
-		log.WithError(err).Warn("ip-masq-agent failed to restore")
+		log.WithError(err).Warn("Failed to restore")
+	}
+	if err := a.Update(); err != nil {
+		log.WithError(err).Warn("Failed to update")
 	}
 
-	manager.UpdateController("ip-masq-agent",
-		controller.ControllerParams{
-			DoFunc: func(ctx context.Context) error {
-				return a.Update()
-			},
-			RunInterval: syncPeriod,
-		},
-	)
+	a.stop = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case event := <-a.watcher.Events:
+				log.Debugf("Received fsnotify event: %+v", event)
+
+				if event.Name != a.configPath {
+					continue
+				}
+
+				switch event.Op {
+				case fsnotify.Create, fsnotify.Write, fsnotify.Chmod, fsnotify.Remove, fsnotify.Rename:
+					if err := a.Update(); err != nil {
+						log.WithError(err).Warn("Failed to update")
+					}
+				default:
+					log.Warnf("Watcher received unknown event: %s. Ignoring.", event)
+				}
+			case err := <-a.watcher.Errors:
+				log.WithError(err).Warn("Watcher received an error")
+			case <-a.stop:
+				log.Info("Stopping ip-masq-agent")
+				return
+			}
+		}
+	}()
+}
+
+// Stop stops the ip-masq-agent goroutine and the watcher.
+func (a *IPMasqAgent) Stop() {
+	close(a.stop)
+	a.watcher.Close()
```

Comment:
Line 152 does not guarantee the channel will be ready first on line 142. Running line 153 without making sure the go routine started on line 122 is terminated might cause a panic ok line 128 since the channel will be closed and return a nil structure.

Reply:
It does not cause panic - if we close the watcher before `<-a.stop` is handled, the watcher sends an error via the `a.watchers.Errors` channel, which is handled in the select.

Also, the `Stop()` method is currently not used.

Reply:
Are you 100% sure? ;-)

Reply:
Yep, while writing the unit tests I accidentally closed the watcher, but not the `a.stop` channel, which resulted in many errors logged by:
```
			case err := <-a.watcher.Errors:
				log.WithError(err).Warn("Watcher received an error")
```

Reply:
But still, this needs to be addressed or there is going to print a read hearing warning in the logs.